xbmc/video/VideoItemArtworkHandler.cpp

```diff
@@ -0,0 +1,472 @@
+/*
+ *  Copyright (C) 2005-2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "VideoItemArtworkHandler.h"
+
+#include "FileItem.h"
+#include "MediaSource.h"
+#include "ServiceBroker.h"
+#include "TextureDatabase.h"
+#include "filesystem/Directory.h"
+#include "guilib/LocalizeStrings.h"
+#include "music/MusicDatabase.h"
+#include "settings/Settings.h"
+#include "settings/SettingsComponent.h"
+#include "utils/FileExtensionProvider.h"
+#include "utils/FileUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
+#include "video/VideoDatabase.h"
+#include "video/VideoInfoScanner.h"
+#include "video/VideoInfoTag.h"
+#include "video/VideoThumbLoader.h"
+#include "video/tags/VideoTagLoaderFFmpeg.h"
+
+using namespace VIDEO;
+using namespace XFILE;
+
+namespace
+{
+//-------------------------------------------------------------------------------------------------
+// CVideoItemArtworkHandler (Generic handler)
+//-------------------------------------------------------------------------------------------------
+
+class CVideoItemArtworkHandler : public IVideoItemArtworkHandler
+{
+public:
+  explicit CVideoItemArtworkHandler(const std::shared_ptr<CFileItem>& item,
+                                    const std::string& artType)
+    : m_item(item), m_artType(artType)
+  {
+  }
+
+  std::string GetCurrentArt() override;
+  std::string GetEmbeddedArt() override;
+  std::vector<std::string> GetRemoteArt() override;
+  std::string GetLocalArt() override;
+
+  std::string GetDefaultIcon() override;
+
+  void AddItemPathToFileBrowserSources(std::vector<CMediaSource>& sources) override;
+
+  void PersistArt(const std::string& art) override;
+
+protected:
+  void AddItemPathStringToFileBrowserSources(std::vector<CMediaSource>& sources,
+                                             const std::string& itemDir,
+                                             const std::string& label);
+
+  const std::shared_ptr<CFileItem> m_item;
+  const std::string m_artType;
+};
+
+std::string CVideoItemArtworkHandler::GetCurrentArt()
+{
+  if (m_artType.empty())
+  {
+    CLog::LogF(LOGERROR, "Art type not set!");
+    return {};
+  }
+
+  std::string currentArt;
+  if (m_item->HasArt(m_artType))
+    currentArt = m_item->GetArt(m_artType);
+  else if (m_item->HasArt("thumb") && (m_artType == "poster" || m_artType == "banner"))
+    currentArt = m_item->GetArt("thumb");
+
+  return currentArt;
+}
+
+std::string CVideoItemArtworkHandler::GetEmbeddedArt()
+{
+  std::string embeddedArt;
+  if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+          CSettings::SETTING_MYVIDEOS_USETAGS))
+  {
+    const std::string fileNameAndPath = m_item->GetVideoInfoTag()->m_strFileNameAndPath;
+    if (URIUtils::HasExtension(fileNameAndPath, ".mkv|.mp4|.avi|.m4v"))
```

Comment:
I'd move this to a utils method

Reply:
Would exactly do you mean? Alwin wanted to come up with a new overload for `URIUtils::HasExtension`, taking a vector of file extensions - and there is PR already for this if I'm not completely mistaken. Alwin would change this code here to use the new overload, given my PR goes in before Alwins.

Reply:
I mean that (personally) I think the handler is interested in knowing if the item supports embedded artwork but should not have to be aware of the specifics (e.g. containers where it's supported).
Moving this to some utils method `SupportsEmbeddedArtwork()` in some utils class helps centralising the code and could help other classes that need to know it as well. Even if not in some utils class moving this to its own method in an anonymous namespace would make it more obvious.
Same goes to the extraction itself - in this case it'd reduce the complexity of the main method.

Reply:
Right, will be changed.